src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -1044,6 +1059,53 @@ - (void)setExpectedValues:(NSArray<NSDictionary<NSString *, id> *> *)values expe
     os_unfair_lock_unlock(&self->_lock);
 }
 
+// MTRDevice should only create read interaction in these cases:
+//   - There is no active subscription
+//   - There is no value in the read cache
+//   - The attribute being read is not reported through a subscription
+//     * Whenever a read interaction returns a result, if there is no read cache entry, then the path would be added to a set that
+//     means "not reported through subscription"
+//     * Any subscription report will clear the path in the set, if exists
+//     * When client calls read, check if the path is in the set, and read-through if true
+- (BOOL)_shouldReadThroughAttributePath:(MTRAttributePath *)attributePath
+{
+    os_unfair_lock_lock(&self->_lock);
+
+    if (_state != MTRDeviceStateReachable) {
+        os_unfair_lock_unlock(&self->_lock);
+        return YES;
+    }
+
+    if (!_readCache[attributePath]) {
+        os_unfair_lock_unlock(&self->_lock);
+        return YES;
+    }
+
+    if ([_nonSubscriptionAttributePaths containsObject:attributePath]) {
+        os_unfair_lock_unlock(&self->_lock);
+        return YES;
+    }
+
+    os_unfair_lock_unlock(&self->_lock);
+    return NO;
+}
+
+- (void)_noteReadInteractionResponseForAttributePath:(MTRAttributePath *)attributePath
+{
+    os_unfair_lock_lock(&self->_lock);
+    if (!_readCache[attributePath]) {
+        [_nonSubscriptionAttributePaths addObject:attributePath];
```

Comment:
I'm not sure the logic works quite right here.

Things that have the C quality are reported in the priming read.  What's not reported is any updates after that point.

So once a subscription is established, we will have values for these things in the `_readCache`, and after that point they will never end up in `_nonSubscriptionAttributePaths`...